### ISSUE TITLE
Defect in zio.Has when combining layers of the same type

### ISSUE THREAD

**Issue by felher** [zio/zio#3334]
Hey folks,

I stumbled upon a runtime exception in `zio.Has` today when doing some experimenting. When you combine a layer with a combination of two layers of the same type, the two layers get erased:

```scala
import zio._
import zio.console._

object Main extends App {
  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] = for {
    _ <- ZIO.accessM[Console with Has[String]](x => putStrLn(x.get[String]))
          .provideLayer(Console.live ++ (ZLayer.succeed("a") ++ ZLayer.succeed("b")))
  } yield 0
}
```

results in a
```
zio.FiberFailure: Fiber failed.
An unchecked error was produced.
java.lang.Error: Defect in zio.Has: Could not find String inside Map(package::Console::Service -> zio.console.package$Console$Service$$anon$1@65735f1f)
```

The problems seems to be that the inferred type of `ZLayer.succeed("a") ++ ZLayer.succeed("b")` is `ZLayer[Any, Nothing, Has[String] with Has[String]]`. But a `Tagged[X with X]`, when summoned, results in a `Tagged[<refinement>]` instead of the expected `Tagged[{X & X}]` or `Tagged[X]`. So when the outer `++` gets applied, `self.unionAll(that.prune)` is called, with `self` being the type map of the left layer and `that` being the type map of the combined right layer. Since the tag of right layer is summoned incorrectly, the `prune` removes everything from the right layer, because the only thing supposed to be there, according to the tag, is `<refinement>`.

I don't think there is a (good) way to fix that within ZIO. I dug around a bit in `zio/izumi-reflect` and think I at least found the source of  the unexpected `Tag[<refinement>]`. I opened an issue here: https://github.com/zio/izumi-reflect/issues/12

Maybe this issue should be kept for the purpose of tracking and pointing users to?

**Comment by neko-kai** [comment]
@felher Could you open a PR adding a test for that in ZIO? This should be fixed once https://github.com/zio/zio/pull/3445 lands

**Comment by felher** [comment]
Sure thing, will do.

**Comment by felher** [comment]
PR opened here: #3506

**Comment by neko-kai** [comment]
Fixed in #3445
Verified in https://github.com/zio/zio/pull/3506
